apps/mysql.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apps
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator"
+	"github.com/GoogleCloudPlatform/ops-agent/confgenerator/fluentbit"
+)
+
+type LoggingProcessorMysqlError struct {
+	confgenerator.ConfigComponent `yaml:",inline"`
+}
+
+func (LoggingProcessorMysqlError) Type() string {
+	return "mysql_error"
+}
+
+func (p LoggingProcessorMysqlError) Components(tag string, uid string) []fluentbit.Component {
+	c := confgenerator.LoggingProcessorParseRegex{
+		// Format documented: https://dev.mysql.com/doc/refman/8.0/en/error-log-format.html
+		// Older versions of mysql should have the same general format, but may not have tid, error_code, subsystem.
+		// Sample Line: 2020-08-06T14:25:02.936146Z 0 [Warning] [MY-010068] [Server] CA certificate /var/mysql/sslinfo/cacert.pem is self signed.
+		// Sample Line: 2020-08-06T14:25:03.109022Z 5 [Note] Event Scheduler: scheduler thread started with id 5
+		Regex: `^(?<time>\d{4}-\d{2}-\d{2}(?:T|\s)\d{2}:\d{2}:\d{2}.\d+(?:Z|[+-]\d{2}:\d{2})?)(?:\s+(?<tid>\d+))?(?:\s+\[(?<level>[^\]]+)])?(?:\s+\[(?<error_code>[^\]]+)])?(?:\s+\[(?<subsystem>[^\]]+)])?\s+(?<message>.*)$`,
```

Comment:
Why do we use `error_code` here but `errorCode` in Apache? Does this match the field name conventions from fluentd? If you're trying to match MySQL's JSON logging format, they use `err_code` and only contain the int, not the whole string with `MY-`.

Reply:
We should aim for consistency, barring some extenuating reason to diverge (e.g. user-expectation, new conventions in upstream etc.).   Also curious what the user-expectation here is and if we should unify w/ what was done for Apache / nginx.

Reply:
`Nginx` set the error code in the httpRequest.status field: https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry#HttpRequest.

`Apache`, as Quentin mentioned, sets an `errorCode` field: https://github.com/GoogleCloudPlatform/ops-agent/pull/213/files/1dff1e9ab954dd85135d108912835c6ce8dcc10a#diff-3934432bf4db161a04479d4460bfbc28eeeab791cbbb49cc166f89e376e0633fR94

The special log fields in the other section of logging config documentation uses camel case:
https://cloud.google.com/stackdriver/docs/solutions/agents/ops-agent/configuration#special-fields

Reply:
Nginx error logs do not parse the error code out into a separate field. They just treat it as part of the message. (I don't think we should extrapolate that here, since MySQL error codes are more well-defined and expected on every log line.)

Nginx access logs fill in `httpRequest.status`, but that's a completely different kind of code, not an error code, so I don't think that applies either.

Reply:
This was an oversight on my part, I was intending to be consistent with the other log parsing examples and stick with camel case. Updating.

Reply:
This has been updated. Resolving the thread.

Reply:
Actually, can we update the screenshot in the PR description as well? (Not a blocker for the PR)